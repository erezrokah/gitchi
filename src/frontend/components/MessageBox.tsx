import * as React from 'react';
import { useState } from 'react';
import styled from 'styled-components';
import { Input } from 'semantic-ui-react';

const StyledInput = styled(Input)`
  min-width: 400px;
`;

interface Props {
  onSendMessage: (message: string) => void;
}

export const MessageBox = (props: Props) => {
  const [message, setMessage] = useState('');

  const handleChange = (e: React.ChangeEvent<HTMLInputElement>) => {
    setMessage(e.target.value);
  };

  const sendMessage = () => {
    props.onSendMessage(message);
    setMessage('');
  };

  const handleKeyPress = (e: React.KeyboardEvent<HTMLInputElement>) => {
    if (e.key === 'Enter') {
      sendMessage();
    }
  };

  return (
    <StyledInput
      fluid
      placeholder="Message..."
      name="message"
      type="text"
      value={message}
      onChange={handleChange}
      onKeyPress={handleKeyPress}
      action={{
        icon: 'send',
        onClick: sendMessage,
        primary: true,
        title: 'Send Message',
      }}
    />
  );
};
