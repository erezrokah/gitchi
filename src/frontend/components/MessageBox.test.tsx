import * as React from 'react';
import { render, fireEvent } from '@testing-library/react';
import { MessageBox } from './MessageBox';

describe('MessageBox', () => {
  test('should match snapshot', () => {
    const { asFragment } = render(<MessageBox onSendMessage={jest.fn()} />);
    expect(asFragment()).toMatchSnapshot();
  });

  test('should update message on input value change', () => {
    const { getByPlaceholderText } = render(
      <MessageBox onSendMessage={jest.fn()} />,
    );

    const message = 'LGTM!';

    const inputElement = getByPlaceholderText('Message...');
    fireEvent.change(inputElement, { target: { value: message } });

    expect(inputElement).toHaveValue(message);
  });

  test('should invoke onSendMessage on button click', () => {
    const onSendMessage = jest.fn();
    const { getByPlaceholderText, getByTestId } = render(
      <MessageBox onSendMessage={onSendMessage} />,
    );

    const message = 'LGTM!';

    fireEvent.change(getByPlaceholderText('Message...'), {
      target: { value: message },
    });
    fireEvent.click(getByTestId('sendMessageButton'));

    expect(onSendMessage).toHaveBeenCalledTimes(1);
    expect(onSendMessage).toHaveBeenCalledWith(message);
  });

  test('should invoke onSendMessage on enter key press', () => {
    const onSendMessage = jest.fn();
    const { getByPlaceholderText } = render(
      <MessageBox onSendMessage={onSendMessage} />,
    );

    const message = 'LGTM!';

    const inputElement = getByPlaceholderText('Message...');
    fireEvent.change(inputElement, {
      target: { value: message },
    });
    fireEvent.keyPress(inputElement, { key: 'Enter', code: 13, charCode: 13 });

    expect(onSendMessage).toHaveBeenCalledTimes(1);
    expect(onSendMessage).toHaveBeenCalledWith(message);
  });

  test('should not invoke onSendMessage on non enter key press', () => {
    const onSendMessage = jest.fn();
    const { getByPlaceholderText } = render(
      <MessageBox onSendMessage={onSendMessage} />,
    );

    const message = 'LGTM!';

    const inputElement = getByPlaceholderText('Message...');
    fireEvent.change(inputElement, {
      target: { value: message },
    });
    fireEvent.keyPress(inputElement, { key: '0', code: 48, charCode: 48 });

    expect(onSendMessage).toHaveBeenCalledTimes(0);
  });
});
