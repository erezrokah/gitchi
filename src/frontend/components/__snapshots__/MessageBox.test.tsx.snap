// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MessageBox should match snapshot 1`] = `
<DocumentFragment>
  .c0 {
  min-width: 400px;
}

<div
    class="ui fluid action input c0"
  >
    <input
      name="message"
      placeholder="Message..."
      type="text"
      value=""
    />
    <button
      class="ui icon primary button"
      data-testid="sendMessageButton"
      title="Send Message"
    >
      <i
        aria-hidden="true"
        class="send icon"
      />
    </button>
  </div>
</DocumentFragment>
`;
